Adding support for secondary interfaces

burhan20 opened the pull request.

abhis2112 committed:
Changes to support telco requirements

The Telco Case requires below requirements:
Multiple pod interfaces
Annotation of secondary interfaces and the CNI plugins used
Supress BGP related information

Commit:
Changes:
1. Show interface name as labels for node to leaf edges.
2. Show multiple links for multiple ifaces.
3. Remove VM_Host from graph(Temp change for baremetal case)
4. See the interface details on the pod-to-node graph.
5. Color code edge for primary iface link and links of secondary ifaces.

Commit:
Changes:
1. Read the nodefabricnetworkattachment, get connection info and provide the new link for worker node to leaf node (different coloring).
2. Remove arrows from "Pod & Nodes" view.
3. Fix bug where the pod to node edges were duplicated on graph
   regeneration.

Commit:
Changes:
1. Read the pod based interface information from nodefabricnetworkattachment CR.
2. Changes for modified structure of nodefabricnetworkattachment CR.
3. Move role and rolebinding to access nodefabricnetworkattachments from
   neo4j template to vkaci template.
4. Modify the queries as per the new pod information from
   nodefabricnetworkattachment CR.

Commit:
Have same coloring scheme for pod to node and node to leaf for primary and secondary links

Commit:
Changes:
1. Add PF, VF for SRIOV ifaces.
2. Fix bug where duplicate nodes were shown.
3. Graph visual changes.

Commit:
Add different colors for primary, secondary and tertiary links

burhan20 committed:
Added three different views for Primary, SRIOV, MacVlan Links

burhan20 committed:
Added changes to update bgp info if ASN is not NONE, fixed code to pass the test cases, modified code to incorporate multiple fabriclink.

abhis2112 commented on app/Dockerfile:
```diff
@@ -2,6 +2,7 @@ FROM python:3.10-slim
 WORKDIR /app
 ENV PATH "$PATH:/app"
 COPY requirements.txt requirements.txt
+RUN echo "deb http://deb.debian.org/debian bookworm main" > /etc/apt/sources.list
```
why this change is needed ?

burhan20 replied:
The build is failing actually. It is not able to find /etc/apt/sources.list. This is not final yet, checking why build is failing

camrossi replied:
Not sure why is happening but started to fail also for my main branch. I just removed the line altogether and works now

burhan20 replied:
reverted these changes

abhis2112 commented on helm/charts/neo4j-vkaci/templates/neo4j-service-account.yaml:
```diff
@@ -36,3 +36,4 @@ roleRef:
   name: {{ .Release.Name }}-service-reader # this must match the name of the Role or ClusterRole you wish to bind to
   apiGroup: rbac.authorization.k8s.io
 {{- end }}
+
```
remove extra line

abhis2112 commented on helm/templates/clusterrole.yaml:
```diff
@@ -29,4 +29,32 @@ subjects:
   kind: ClusterRole
   name: {{ include "vkaci.fullname" . }}-pods-list
+  apiGroup: rbac.authorization.k8s.io
+---
```
ideally should be a separate file naming role.yaml which includes both role and rolebinding  as here in clusterrole.yaml

abhis2112 commented on unit_test.py:
```diff
@@ -223,13 +224,13 @@ def test_valid_topology(self):
-                                          'labels': {'app': 'redis'}, 'mac': 'MOCKMO1C'}},
-                    'services': {'appx': [{'name': 'example service', 'cluster_ip': '192.168.25.5', 'external_i_ps': ['192.168.5.1'],'load_balancer_ip': '192.168.5.2','ns':'appx',
-                                           'labels': {'app': 'guestbook'}}]}}
+                                          'labels': {'app': 'redis'}, 'node_leaf_sec_iface_conn': [], 'node_pod_sec_iface_conn': [], 'node_leaf_ter_iface_conn': [], 'node_pod_ter_iface_conn': [], 'node_leaf_all_iface_conn': [], 'mac': 'MOCKMO1C'}},
```
we have added null values everywhere . Is this correct? Should be some value right

abhis2112 replied:
same comment wherever we are adding null values

burhan20 replied:
We have added these additional parameters in our structure. We had to pass these parameters to the already present testcases as well otherwise the testcases would fail.

burhan20 committed:
Wrote the role and rolebindings for nfna in separate file

burhan20 committed:
removed extra line

camrossi commented on app/graph.py:
```diff
@@ -36,7 +37,7 @@ def __init__(self, dict_env:dict = None):
             self.apic_ip = []
 
-        self.aciMetaFilePath = self.enviro().get("ACI_META_FILE", "/app/aci-meta/aci-meta.json")
+        self.aciMetaFilePath = self.enviro().get("ACI_META_FILE", "/root/.aci-meta/aci-meta.json")
```
I had to change from `/root` to `/app` to make it work across OpenShift and Upstream Kubernetes and I fear this might break things again. Need to test it in my own environment

camrossi replied:
Indeed this breaks compatibility with Upstream K8s
```
pi@pi-k8s-1:~$ kubectl  logs vkaci-7548c967f8-slds6 -c init-vkaci
2023-06-15 17:05:40,284 __main__ INFO Loading ACI Metadata
Traceback (most recent call last):
  File "/app/init.py", line 29, in <module>
    file = open('/root/.aci-meta/aci-meta.json','wb')
FileNotFoundError: [Errno 2] No such file or directory: '/root/.aci-meta/aci-meta.json'
```

Just revert this back to `/app/aci-meta/aci-meta.json`

camrossi commented on app/graph.py:
```diff
@@ -470,23 +487,108 @@ def update(self):
+            except Exception as e:
+                logger.error(f"Error processing pod: {i.metadata.name}. Error: {str(e)}")
+
+        cr = self.custom_obj.list_namespaced_custom_object(group="aci.fabricattachment", version="v1", namespace="aci-containers-system", plural="nodefabricnetworkattachments")
```
This resource might not exists, at the moment we crash if is not present:
```
kubernetes.client.exceptions.ApiException: (404)
Reason: Not Found
HTTP response headers: HTTPHeaderDict({'Audit-Id': '3d0a3afb-ef4a-4163-9341-025e45d0ca3c', 'Cache-Control': 'no-cache, private', 'Content-Type': 'text/plain; charset=utf-8', 'X-Content-Type-Options': 'nosniff', 'X-Kubernetes-Pf-Flowschema-Uid': 'c3d0b261-bb0f-4204-ba5a-c013697f1875', 'X-Kubernetes-Pf-Prioritylevel-Uid': 'e2fd79d1-4c6d-4bf6-86a8-eea82de11102', 'Date': 'Thu, 15 Jun 2023 17:43:07 GMT', 'Content-Length': '19'})
HTTP response body: 404 page not found
```
we should handle this gracefully as not all cluster where VKACI runs have SR-IOV

abhis2112 replied:
@burhan20 i think with latest changes we expect CRD to be present which might not be the case . We need to add try/except  block. Something like below:
```
     try:
            self.custom_obj.get_namespaced_custom_object(group="aci.fabricattachment", version="v1", namespace="aci-containers-system", plural="nodefabricnetworkattachments")
            resource_exists = True
        except client.exceptions.ApiException as e:
            if e.status == 404:
                resource_exists = False
            else:
                raise

        # If the custom resource exists, list objects
        if resource_exists:
            cr = self.custom_obj.list_namespaced_custom_object(group="aci.fabricattachment", version="v1",
                                                               namespace="aci-containers-system",
                                                               plural="nodefabricnetworkattachments")
            if cr.get("items"):
                for nodeName in self.topology['nodes']:
                    try:
                        for i in cr.get("items"):
                            aciTopology = i["spec"].get("aciTopology")
                            if aciTopology is not None and i["spec"]["nodeName"] == nodeName:
                                for iface_name, link in i["spec"]["aciTopology"].items():
                                    if "sriov" in i["metadata"]["name"]:
                                        iface_name = "PF-" + iface_name
                                    fabricLinks = link.get("fabricLink", [])
                                    for fabricLink in fabricLinks:
                                        fabricLinkSplit = fabricLink.split("/")
                                        switch_name = fabricLinkSplit[2].replace("node", "leaf")
                                        switch_interface = fabricLink[fabricLink.rfind('[') + 1: fabricLink.rfind(']')]
                                        if "sriov" in i["metadata"]["name"]:
                                            self.sriov = True
                                            self.topology['nodes'][nodeName]['node_leaf_sec_iface_conn'].append({
                                                'switch_name': switch_name,
                                                'switch_interface': switch_interface,
                                                'node_iface': iface_name
                                            })
                                        else:
                                            self.macvlan = True
                                            self.topology['nodes'][nodeName]['node_leaf_ter_iface_conn'].append({
                                                'switch_name': switch_name,
                                                'switch_interface': switch_interface,
                                                'node_iface': iface_name
                                            })

                                    pods = link.get("pods", [])
                                    for pod in pods:
                                        node_iface = pod.get("localIface")
                                        if "sriov" in i["metadata"]["name"]:
                                            node_iface = "VF-" + node_iface
                                        pod_name = pod.get("podRef")["name"]
                                        network_ref = i["spec"].get("networkRef")
                                        node_network = network_ref["name"]
                                        if "sriov" in i["metadata"]["name"]:
                                            self.topology['nodes'][nodeName]['node_pod_sec_iface_conn'].append({
                                                'node_iface': node_iface,
                                                'pod_name': pod_name,
                                                'node_network': node_network,
                                                'pod_iface': self.topology['nodes'][nodeName]['pods'][pod_name][
                                                    'other_ifaces'].get(node_network, "")
                                            })
                                        else:
                                            self.topology['nodes'][nodeName]['node_pod_ter_iface_conn'].append({
                                                'node_iface': node_iface,
                                                'pod_name': pod_name,
                                                'node_network': node_network,
                                                'pod_iface': self.topology['nodes'][nodeName]['pods'][pod_name][
                                                    'other_ifaces'].get(node_network, "")
                                            })

                        self.topology['nodes'][nodeName]['node_leaf_all_iface_conn'].extend(
                            self.topology['nodes'][nodeName]['node_leaf_sec_iface_conn'])
                        self.topology['nodes'][nodeName]['node_leaf_all_iface_conn'].extend(
                            self.topology['nodes'][nodeName]['node_leaf_ter_iface_conn'])

                    except Exception as e:
                        logger.error(f"Error processing node: {nodeName}. Error: {str(e)}")

camrossi replied:
This works fine in my lab:

```
        try:
            cr = self.custom_obj.list_namespaced_custom_object(group="aci.fabricattachment", version="v1", namespace="aci-containers-system", plural="nodefabricnetworkattachments")
                    ###all the code###
                    except Exception as e:
                        logger.error(f"Error processing node: {nodeName}. Error: {str(e)}")
        except Exception as e:
            if e.status == 404:
                logger.info(f"CRD nodefabricnetworkattachments not detected, SR-IOV/MACVLAN topology support is disabled")
            else:
                logger.error(f"Unexpected error processing list_namespaced_custom_object. Error: {str(e)}")

camrossi commented on app/template/assets/cui-2.0.0/js/vkaci.js:
```diff
@@ -16,118 +13,160 @@ function neo_viz_config(showPodName, container, cypher, seed = null) {
+                iterations: 400,
+                fit: true,
+            },
+            dragView: true
```
In my environment this leads to a crash:
```
Unknown option detected: "dragView" in 

options = {
  physics: {
    dragView
  }
}
```

Not sure why, do you use a different neo4j version ?

camrossi commented on app/graph.py:
```diff
@@ -626,30 +728,78 @@ def __init__(self, env: VkaciEnvVariables, topology: VkaciBuilTopology) -> None:
 
     # Build query.
     query = """
```
I think all the query need to be re-written to account the fact that not all cluster will have the `fabricattachment` CR 
Currently if I run this code on a cluster without the CR no visualisation happens

camrossi commented on app/template/index.html:
```diff
@@ -136,6 +136,9 @@
                               <a id="WithoutBgpPeers" onclick="draw_without_bgp_peers()">Without Bgp Peers</a>
                               <a id="WithoutPods" onclick="draw_without_pods()">Without Pods</a>
                               <a id="OnlyBgpPeers" onclick="draw_only_bgp_peers()">BGP Peers</a>
+                              <a id="OnlyPrimarylinks" onclick="draw_only_primary_links()">Primary Links</a>
```
These new 3 views should only be displayed if the CR `fabricattachment` is present. 
We need to support also clusters without this CR as we have customers using this tool unrelated to SR-IOV

burhan20 committed:
Added back VM host in between node and leaf in non openshift case

burhan20 committed:
Filtered the UI menu based on the condition whether it is openshift case or not

camrossi commented on app/template/index.html:
```diff
@@ -133,9 +133,18 @@
                            <div id="selected_views" class="dropdown__menu dropdown__menu--openleft">
                               <a id="PodsAndNodes" onclick="draw_pods_and_nodes()">Pods & Nodes</a>
-                              <a id="WithoutBgpPeers" onclick="draw_without_bgp_peers()">Without Bgp Peers</a>
+                              {% if not openshift %}
```
I fear the logic is not correct or at minimum to wording is wrong.
We already support OpenShift (without SR-IOV or MacVlan)

What you are adding is support for SR-IOV and MacVALN so I think what you should check is if the new CR is present then you show the menus.
Having a `if not openshift` kinda check is confusing

burhan20 replied:
Cam, as per the latest changes, we are showing Primary, SRIOV, Macvlan links in dropdown when CR is present.
However, BGP Peers and Without BGP Peers menu we are showing if not openshift case.
Is it the right way to do? Could you please suggest what needs to be done.

burhan20 committed:
Modified code to filter Primary, SR-IOV, MacVlan links based on presence of nfna cr

abhis2112 commented on app/graph.py:
```diff
@@ -470,23 +488,111 @@ def update(self):
+
+        cr = self.custom_obj.list_namespaced_custom_object(group="aci.fabricattachment", version="v1", namespace="aci-containers-system", plural="nodefabricnetworkattachments")
+
+        if cr:
```
you need to check if len(cr) == 0 as it is a list

abhis2112 replied:
I am not able to view any topology even when it is false. Can you check the logic again here

burhan20 committed:
Fixed the issue of pods not showing up when CR are not present

burhan20 committed:
Changes:
1.  Fixed query incase node_leaf_all_iface_conn is empty to show worker
    node and its pods
2.  Removed the VM host between the node(s) and leaf(s) for baremetal
    case.
3.  Fixed the bug of links between node(s) and leaf(s) getting same
    label for baremetal case
4.  Fixed the issue of graph not updating on regenerating the
    graph after deleting the CR.

burhan20 committed:
Fixed sriov menu not showing graph when pod is not present and fixed redundancy in connected_switch_ifaces

burhan20 committed:
Modified Views dropdown: Showing SRIOV/MacVlan option only when respective cr is present

abhis2112 commented on app/graph.py:
```diff
@@ -470,23 +493,112 @@ def update(self):
+                                    fabricLinkSplit = fabricLink.split("/")
+                                    switch_name = fabricLinkSplit[2].replace("node", "leaf")
+                                    switch_interface = fabricLink[fabricLink.rfind('[') + 1: fabricLink.rfind(']')]
+                                    if "sriov" in i["metadata"]["name"]:
```
we need to come up with a more robust solution that how we want to check for the type. The name might not be enough. Will discuss in the group want we want to select a type - a name, or an annotation

burhan20 committed:
Updated base image to python:3.10-slim-bullseye

burhan20 committed:
Optimized code

burhan20 committed:
Addressed comments:

1. Removed labels from the queries
2. Wrapped the CR fetch in try except block
3. Removed dragview from neovis
4. Removed unnecessary commas while showing interfaces name
5. Added optional match for primary link query

burhan20 committed:
Reverted volume mount changes, fixed RunAsUser value, changed variable name for filtering bgp

burhan20 committed:
Added namespace for serviceaccount

burhan20 committed:
Fixed query in view all and without pods view to show nodes even if pod is not present

abhis2112 commented on helm/charts/neo4j-vkaci/values.yaml:
```diff
@@ -214,7 +214,7 @@ config:
 # securityContext defines privilege and access control settings for a Pod or Container. Making sure that we dont run Neo4j as root user.
 securityContext:
   runAsNonRoot: true
-  runAsUser: 1000700000
```
is this change required

burhan20 replied:
I will revert this. This value is different for each setup

abhis2112 commented on helm/values.yaml:
```diff
@@ -45,7 +45,7 @@ podSecurityContext: {}
 #This ire required for OpenShift to run the container as non-root 
 securityContext:
   runAsNonRoot: true
-  runAsUser: 1000700000
```
same here

burhan20 committed:
Added back VM host in topology and reverted the changed to original code for showing switch-VM interface label

burhan20 committed:
Removed the VM and fixed the labels between node and leaf
